presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergPageSourceProvider.java

```diff
@@ -0,0 +1,667 @@
+            for (int columnIndex = 0; columnIndex < regularColumns.size(); columnIndex++) {
+                IcebergColumnHandle column = regularColumns.get(columnIndex);
+                namesBuilder.add(column.getName());
+                org.apache.parquet.schema.Type parquetField = parquetFields.get(columnIndex);
+
+                Type prestoType = column.getType();
+
+                prestoTypes.add(prestoType);
+
+                if (parquetField == null) {
+                    internalFields.add(Optional.empty());
+                }
+                else {
+                    internalFields.add(constructField(column.getType(), messageColumnIO.getChild(parquetField.getName())));
+                }
+            }
+
+            return new ParquetPageSource(parquetReader, prestoTypes.build(), internalFields.build(), namesBuilder.build(), new RuntimeStats());
+        }
+        catch (Exception e) {
+            try {
+                if (dataSource != null) {
+                    dataSource.close();
+                }
+            }
+            catch (IOException ignored) {
+            }
+            if (e instanceof PrestoException) {
+                throw (PrestoException) e;
+            }
+            String message = format("Error opening Iceberg split %s (offset=%s, length=%s): %s", path, start, length, e.getMessage());
+
+            if (e instanceof ParquetCorruptionException) {
+                throw new PrestoException(ICEBERG_BAD_DATA, message, e);
+            }
+
+            if (e instanceof BlockMissingException) {
+                throw new PrestoException(ICEBERG_MISSING_DATA, message, e);
+            }
+            throw new PrestoException(ICEBERG_CANNOT_OPEN_SPLIT, message, e);
+        }
+    }
+
+    private static TupleDomain<ColumnDescriptor> getParquetTupleDomain(Map<List<String>, RichColumnDescriptor> descriptorsByPath, TupleDomain<IcebergColumnHandle> effectivePredicate)
+    {
+        if (effectivePredicate.isNone()) {
+            return TupleDomain.none();
+        }
+
+        ImmutableMap.Builder<ColumnDescriptor, Domain> predicate = ImmutableMap.builder();
+        effectivePredicate.getDomains().get().forEach((columnHandle, domain) -> {
+            String baseType = columnHandle.getType().getTypeSignature().getBase();
+            // skip looking up predicates for complex types as Parquet only stores stats for primitives
+            if (!baseType.equals(StandardTypes.MAP) && !baseType.equals(StandardTypes.ARRAY) && !baseType.equals(StandardTypes.ROW)) {
+                RichColumnDescriptor descriptor = descriptorsByPath.get(ImmutableList.of(columnHandle.getName()));
+                if (descriptor != null) {
+                    predicate.put(descriptor, domain);
+                }
+            }
+        });
+        return TupleDomain.withColumnDomains(predicate.build());
+    }
+
+    private static ConnectorPageSource createBatchOrcPageSource(
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 17. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.